### ISSUE TITLE
Storage class cannot be JSON serialized

### ISSUE THREAD

**Issue by girardinsamuel** [codingjoe/django-stdimage#252]
I am using async rendering detailed here https://github.com/codingjoe/django-stdimage#async-image-processing

When using S3 storage, the celery task receives a storage argument which is the storage instance used for the current file.
Celery is raising an error as task arguments should be "JSON serializable" and the storage instance cannot be serialized as JSON...

Do you see a workaround ?

Thank you very much !

**Comment by girardinsamuel** [comment]
The exact exception is `Object of type PrivateS3Storage is not JSON serializable`
where `PrivateS3Storage` is a `S3Boto3Storage` class.

If I am not giving the storage as argument and instantiating the storage myself inside the task I get another error. Here is the pseudo code adapted from the doc example :

```
from celery import shared_task

from stdimage.utils import render_variations

from my_app.storages import PrivateS3Storage

@shared_task
def process_photo_image(file_name, variations):
    storage = PrivateS3Storage()
    render_variations(file_name, variations, replace=True, storage=storage)
```
from django.db import models
from stdimage.models import StdImageField

from .tasks import process_photo_image
from my_app.storages import PrivateS3Storage


def image_processor(file_name, variations, storage):
    # here I removed storage instance argument to task as it can't be serialized
    process_photo_image.delay(file_name, variations)
    return False

class AsyncImageModel(models.Model):
    image = StdImageField(
        upload_to='path/to/file/',
        storage=PrivateS3Storage()
        render_variations=image_processor
    )
```

**Comment by girardinsamuel** [comment]
I just come accross #173... I will try what you propose !

**Comment by girardinsamuel** [comment]
It's working quite easily ! Here is the approach I took :

## TLDR;
In the `render_variations` callable: parse the storage instance received in the args a storage path.
Later, in the task re-instantiate the storage instance.

## Code example

```python
# tasks.py
from celery import shared_task
from stdimage.utils import render_variations


@shared_task
def process_user_avatar(file_name, variations, storage_path):
    # reconstruct storage instance as Django would do
    from django.core.files.storage import get_storage_class

    class SelectedStorage(LazyObject):
        def _setup(self):
            self._wrapped = get_storage_class(storage_path)()

    storage = SelectedStorage()
    render_variations(file_name, variations, replace=True, storage=storage)

```

```
# models.py

 def image_processor(file_name, variations, storage):
        # convert storage instance to storage_path to provide as task argument
        storage_path = f"{storage.__class__.__module__}.{storage.__class__.__name__}"
        process_user_avatar.delay(file_name, variations, storage_path)
        return False  # prevent default rendering


class AsyncImageModel(models.Model):
    image = StdImageField(
        upload_to='path/to/file/',
        storage=PrivateS3Storage()  # here you can specify storage or not and the default will be used
        render_variations=image_processor
    )
```
